ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -1112,25 +1115,25 @@ void handleBody(InputStream payload, MessageMetadata messageMetadata, MessageInf
         ByteBuffer encryptionKey = messageMetadata == null ? null : messageMetadata.getEncryptionKey();
         ByteBuf chunkBuf = blobData.content();
         this.isChunkCompressed = blobData.isCompressed();
+        // We are going to send this chunkBuf later back to the router, so increase the counter here
+        chunkBuf.retain();
 
-        try {
-          boolean launchedJob = maybeLaunchCryptoJob(chunkBuf, null, encryptionKey, chunkBlobId);
-          if (!launchedJob) {
-            if (chunkIndex == 0) {
-              routerMetrics.getFirstDataChunkLatencyMs.update(System.currentTimeMillis() - submissionTimeMs);
-            }
-            routerMetrics.getDataChunkLatencyMs.update(System.currentTimeMillis() - initializedTimeMs);
-            ByteBuf decompressedContent = decompressContent(chunkBuf.retainedDuplicate());
-            if (decompressedContent != null) {
-              chunkIndexToBuf.put(chunkIndex, filterChunkToRange(decompressedContent));
-              numChunksRetrieved.incrementAndGet();
-            }
+        boolean launchedJob = maybeLaunchCryptoJob(chunkBuf, null, encryptionKey, chunkBlobId);
+        if (!launchedJob) {
+          if (chunkIndex == 0) {
+            routerMetrics.getFirstDataChunkLatencyMs.update(System.currentTimeMillis() - submissionTimeMs);
+          }
+          routerMetrics.getDataChunkLatencyMs.update(System.currentTimeMillis() - initializedTimeMs);
+          ByteBuf decompressedContent = decompressContent(chunkBuf);
+          // If the decompressed content is valid, don't use chunkBuf anymore, so we release it.
+          // If the decompressed content is null, which means the operation failed, just release the chunkBuf.
+          chunkBuf.release();
```

Comment:
Just a thought. Since `decompressContent()` method is the one who is creating a new ByteBuf for storing decompressed content, would it be better if it releases the passed input ByteBuf?